torch_xla/_internal/gpu.py

```diff
@@ -14,7 +15,8 @@ def num_local_processes() -> int:
   """
   assert xenv.GPU_NUM_DEVICES in os.environ, \
       "Must set `GPU_NUM_DEVICES` environment variable to use the PjRt GPU client"
-  return int(os.environ[xenv.GPU_NUM_DEVICES])
+  os.environ[xenv.WORLD_SIZE] = os.environ[xenv.GPU_NUM_DEVICES]
```

Comment:
Is this right? We'll clobber the torchrun-set world size. Also wondering if we need to keep `GPU_NUM_DEVICES` in the first place

Reply:
This change is mainly to make single-host case work with `spawn` (e.g. the tests in pytorch/xla/test/pjrt/test_runtime_gpu.py in which we use `spawn`). To provide a similar UX as PyTorch, we should still support `spawn` for the single host case (fwiw, PyTorch supports it as https://screenshot.googleplex.com/7nKD68dXNUUskF7). But I like your idea of replacing "GPU_NUM_DEVICES" with "WORLD_SIZE".

Also, torchrun doesn't invoke `spawn` and in turn this function so it wouldn't clobber the torchrun-set world size.

So how about I replace `GPU_NUM_DEVICES ` with `WORLD_SIZE `?

Reply:
`WORLD_SIZE` isn't quite right here, since this function returns the expected number of _local_ processes. In `torchrun`, that's set as `LOCAL_WORLD_SIZE`. In `xmp.spawn`, we had to set it as something different like `PJRT_LOCAL_WORLD_SIZE` because `LOCAL_WORLD_SIZE` caused some issue with a third party package.

So maybe `os.environ.get('LOCAL_WORLD_SIZE') or os.environ.get('PJRT_LOCAL_WORLD_SIZE')`? It's clunky, but it covers both cases.

To use a subset of local GPUs with `xmp.spawn`, a user could set then `LOCAL_WORLD_SIZE` themselves instead of `GPU_NUM_DEVICES`.

Reply:
Agreed that `WORLD_SIZE` is confusing. OTOH, `WORLD_SIZE` is what PyTorch single-node training uses: https://screenshot.googleplex.com/BH27HAYTpbNU4KA, if we want to stay closer to PyTorch.

I think clarity is more important here. How do you think we replace `GPU_NUM_DEVICES` with `PJRT_LOCAL_WORLD_SIZE` (or something like `PT_XLA_LOCAL_WORLD_SIZE` if we don't want to leak the underlying implementation detail), so the way we run single-host-multi-GPU `GPU_NUM_DEVICES=4 python3 xla/test/test_train_mp_imagenet.py ...` turns to `PJRT_LOCAL_WORLD_SIZE=4 python3 xla/test/test_train_mp_imagenet.p ..`?  @will-cromar @jonb377

Reply:
I think it's okay to rely on the `LOCAL_WORLD_SIZE` env var here. We know torchrun will set it for sure, and the manual single-host-multi-GPU case can become `PJRT_DEVICE=GPU LOCAL_WORLD_SIZE=4 python script.py`*. We can get rid of `GPU_NUM_DEVICES`.

\* As a follow up, I would like to implement more automatic configuration like we have with TPUs so users don't have to set anything in the default case

We definitely don't want to override `torchrun`'s settings here.

I forgot that `PJRT_LOCAL_WORLD_SIZE` is set after this function is called (and probably just set to the output of this function). So we can ignore that variable here.

Reply:
Yeah, I used `LOCAL_WORLD_SIZE ` here.

I'll replace `GPU_NUM_DEVICES` with `LOCAL_WORLD_SIZE` in a follow-up PR.